src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs

```diff
@@ -671,28 +672,51 @@ private async Task EnsureNuGetAndVsProjectAdapterCacheAsync()
             {
                 await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
 
-                if (!_cacheInitialized && await IsSolutionOpenAsync())
+                IVsSolution ivsSolution = await _asyncVSSolution.GetValueAsync();
+                if (!_cacheInitialized && IsSolutionOpenFromVSSolution(ivsSolution))
                 {
                     try
                     {
-                        var dte = await _dte.GetValueAsync();
-
-                        foreach (var project in await EnvDTESolutionUtility.GetAllEnvDTEProjectsAsync(dte))
+                        if (await _featureFlagService.IsFeatureEnabledAsync(NuGetFeatureFlagConstants.NuGetSolutionCacheInitilization))
                         {
-                            try
+                            foreach (var hierarchy in VsHierarchyUtility.GetAllLoadedProjects(ivsSolution))
                             {
-                                var vsProjectAdapter = await _vsProjectAdapterProvider.CreateAdapterForFullyLoadedProjectAsync(project);
-                                await AddVsProjectAdapterToCacheAsync(vsProjectAdapter);
+                                try
+                                {
+                                    var vsProjectAdapter = await _vsProjectAdapterProvider.CreateAdapterForFullyLoadedProjectAsync(hierarchy);
+                                    await AddVsProjectAdapterToCacheAsync(vsProjectAdapter);
+                                }
+                                catch (Exception e)
+                                {
+                                    // Ignore failed projects.
+                                    _logger.LogWarning($"The project {VsHierarchyUtility.GetProjectPath(hierarchy)} failed to initialize as a NuGet project.");
+                                    _logger.LogError(e.ToString());
+                                }
+
+                                // Consider that the cache is initialized only when there are any projects to add.
+                                _cacheInitialized = true;
                             }
-                            catch (Exception e)
+                        }
+                        {
```

Comment:
I'm pretty sure when the feature flag is enabled that you don't want to do both IVsHeirarchy and also DTE adapter creation, so I think you're missing an `else` here.

If I'm wrong and you do want to do both, I think a comment should be added, because it really looks like a bug.

Reply:
I'm missing an else. Fixed.